### PR TITLE
[Example] DAGNN

### PR DESCRIPTION
**Pull request by lt610**
## DAGNN
DGL implementation of DAGNN, which is the model in paper [Towards Deeper Graph Neural Networks](https://arxiv.org/abs/2007.09296).

### CODE REVIEW

**Comment by mufeili** [comment]
Add an entry in the indexing [here](https://github.com/dmlc/dgl/blob/master/examples/README.md).

File: examples/pytorch/dagnn/README.md [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+### Performance
+
+#### On Cora, Citeseer and Pubmed
+##### Accuracy
+| Dataset | Cora | Citeseer | Pubmed |
+| :-: | :-: | :-: | :-: |
+| DAGNN | 84.4 ± 0.5 | 73.3 ± 0.6 | 80.5 ± 0.5 |
+| DGL | 84.3 ± 0.6 | 73.5 ± 0.8 | 80.5 ± 0.3 |
+
+#### On Arxiv
```
Comment (mufeili) [inline examples/pytorch/dagnn/README.md:12]: Is this ogbn-arxiv? If so, I guess it's better to use that term. Also you can directly merge the two tables.

**Comment by lt610** [comment]
Yes, it's ogbn-arxiv. Sorry, I will revise and complete the readme doc.




------------------&nbsp;原始邮件&nbsp;------------------
发件人: "Mufei Li"<notifications@github.com&gt;; 
发送时间: 2021年1月19日(星期二) 下午3:09
收件人: "dmlc/dgl"<dgl@noreply.github.com&gt;; 
抄送: "刘唐"<1421681158@qq.com&gt;; "Author"<author@noreply.github.com&gt;; 
主题: Re: [dmlc/dgl] dagnn (#2545)





 
@mufeili commented on this pull request.
 
 
In examples/pytorch/dagnn/README.md:
 &gt; @@ -0,0 +1,34 @@ +# DAGNN + +### Performance + +#### On Cora, Citeseer and Pubmed +##### Accuracy +| Dataset | Cora | Citeseer | Pubmed | +| :-: | :-: | :-: | :-: | +| DAGNN | 84.4 ± 0.5 | 73.3 ± 0.6 | 80.5 ± 0.5 | +| DGL | 84.3 ± 0.6 | 73.5 ± 0.8 | 80.5 ± 0.3 | + +#### On Arxiv  
Is this ogbn-arxiv? If so, I guess it's better to use that term. Also you can directly merge the two tables.

File: examples/pytorch/dagnn/README.md [hunk 2]
```diff
@@ -0,0 +1,34 @@
+# DAGNN
+
```
Comment (mufeili) [inline examples/pytorch/dagnn/README.md:2]: Add the commands for running the scripts and reproducing the results.

File: examples/pytorch/dagnn/main.py [hunk 3]
```diff
@@ -0,0 +1,272 @@
...
+    return best_acc
+
+
+if __name__ == "__main__":
+    """
+        DAGNN Model Hyperparameters
+    """
+    parser = argparse.ArgumentParser(description='DAGNN')
+    # data source params
+    parser.add_argument('--dataset', type=str, default='Cora', help='Name of dataset.')
```
Comment (mufeili) [inline examples/pytorch/dagnn/main.py:240]: Use `choices` to specify the datasets allowed.

File: examples/pytorch/dagnn/README.md [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+### Performance
+
```
Comment (mufeili) [inline examples/pytorch/dagnn/README.md:4]: Add the number of runs used.

File: examples/pytorch/dagnn/main.py [hunk 5]
```diff
@@ -0,0 +1,272 @@
...
+    # cuda params
+    parser.add_argument('--gpu', type=int, default=0, help='GPU index. Default: -1, using CPU.')
+    # training params
+    parser.add_argument('--runs', type=int, default=1, help='Training runs.')
+    parser.add_argument('--epochs', type=int, default=1500, help='Training epochs.')
+    parser.add_argument('--early-stopping', type=int, default=100, help='Patient epochs to wait before early stopping.')
+    parser.add_argument('--lr', type=float, default=0.01, help='Learning rate.')
+    parser.add_argument('--lamb', type=float, default=0.005, help='L2 reg.')
+    # model params
+    parser.add_argument('--k', type=int, default=10, help='Number of propagation.')
```
Comment (mufeili) [inline examples/pytorch/dagnn/main.py:250]: "Number of propagation." -> "Number of propagation layers." or simply "Depth"

File: examples/pytorch/dagnn/main.py [hunk 6]
```diff
@@ -0,0 +1,272 @@
...
+    parser.add_argument("--hid-dim", type=int, default=64, help='Hidden layer dimensionalities.')
+    parser.add_argument('--batchnorm', action='store_true', default=False, help="batchnorm")
+    parser.add_argument('--dropout', type=float, default=0.8, help='dropout')
+    parser.add_argument('--drop_bef', action='store_true', default=True, help='Location of dropout')
```
Comment (mufeili) [inline examples/pytorch/dagnn/main.py:254]: action='store_false'?

File: examples/pytorch/dagnn/main.py [hunk 7]
```diff
@@ -0,0 +1,272 @@
...
+
+    # Step 1: Prepare graph data and retrieve train/validation/test index ============================= #
+    # Load from DGL dataset
+    if args.dataset == 'Cora':
+        dataset = CoraGraphDataset()
+    elif args.dataset == 'Citeseer':
+        dataset = CiteseerGraphDataset()
+    elif args.dataset == 'Pubmed':
+        dataset = PubmedGraphDataset()
+    else:
```
Comment (mufeili) [inline examples/pytorch/dagnn/main.py:141]: If currently the scripts only support `Cora`, `Citeseer` and `Pubmed`, then you don't need to include the results for other datasets in README.

File: examples/pytorch/dagnn/utils.py [hunk 8]
```diff
@@ -0,0 +1,29 @@
...
+        results = ()
+        for idx in idxs:
+            loss = F.cross_entropy(logits[idx], labels[idx])
+            acc = torch.sum(logits[idx].argmax(dim=1) == labels[idx]).item() / len(idx)
+            results += (loss, acc)
+    return results
+
+
+def generate_random_seeds(seed, nums):
+    random.seed(seed)
```
Comment (mufeili) [inline examples/pytorch/dagnn/utils.py:20]: `random.seed(seed)` is redundant.

File: examples/pytorch/dagnn/utils.py [hunk 9]
```diff
@@ -0,0 +1,29 @@
...
+    return [random.randint(1, 999999999) for _ in range(nums)]
+
+
+def set_random_state(seed):
+    random.seed(seed)
+    np.random.seed(seed)
+    torch.manual_seed(seed)
+    torch.cuda.manual_seed_all(seed)
```
Comment (mufeili) [inline examples/pytorch/dagnn/utils.py:28]: Is this valid if cuda is not available?

File: examples/pytorch/dagnn/main.py [hunk 10]
```diff
@@ -0,0 +1,272 @@
...
+        raise ValueError('Dataset {} is invalid.'.format(args.dataset))
+
+    graph = dataset[0]
+    graph = graph.remove_self_loop().add_self_loop()
```
Comment (mufeili) [inline examples/pytorch/dagnn/main.py:145]: Explain why we want this operation in comments.

File: examples/pytorch/dagnn/main.py [hunk 11]
```diff
@@ -0,0 +1,272 @@
...
+                  batchnorm=args.batchnorm,
+                  dropout=args.dropout,
+                  drop_bef=args.drop_bef)
+    model = model.to(device)
+
+    # Step 3: Create training components ===================================================== #
+    loss_fn = F.cross_entropy
+    opt = torch.optim.Adam(model.parameters(), lr=args.lr, weight_decay=args.lamb)
+
+    # Step 4: training epoches =============================================================== #
```
Comment (mufeili) [inline examples/pytorch/dagnn/main.py:189]: "epoches" -> "epochs"

File: examples/pytorch/dagnn/main.py [hunk 12]
```diff
@@ -0,0 +1,272 @@
...
+
+        # compute loss
+        train_loss = loss_fn(logits[train_idx], labels[train_idx])
+
+        # backward
+        opt.zero_grad()
+        train_loss.backward()
+        opt.step()
+
+        # Validation using a full graph
```
Comment (mufeili) [inline examples/pytorch/dagnn/main.py:208]: I guess you can remove this comment as you are also training over the full graph.

File: examples/pytorch/dagnn/main.py [hunk 13]
```diff
@@ -0,0 +1,272 @@
...
+        train_loss, train_acc, valid_loss, valid_acc, test_loss, test_acc = evaluate(model, graph, feats, labels,
+                                                                                     (train_idx, val_idx, test_idx))
+
+        # Print out performance
+        epochs.set_description('Train Acc {:.4f} | Train Loss {:.4f} | Val Acc {:.4f} | Val loss {:.4f}'.format(
+            train_acc, train_loss.item(), valid_acc, valid_loss.item()))
+
+        if loss is None:
+            loss = valid_loss
+            best_acc = test_acc
```
Comment (mufeili) [inline examples/pytorch/dagnn/main.py:218]: Alternatively, you can set the initial `loss` to be `float('inf')` and initial `best_acc` to be `0`.

File: examples/pytorch/dagnn/main.py [hunk 14]
```diff
@@ -0,0 +1,272 @@
...
+                 dropout=0,
+                 drop_bef=True):
+        super(DAGNN, self).__init__()
+        self.mlp = nn.ModuleList()
+        self.mlp.append(MLPLayer(in_dim=in_dim, out_dim=hid_dim, bias=bias,
+                                 activation=activation, batchnorm=batchnorm,
+                                 dropout=dropout, drop_bef=drop_bef))
+        self.mlp.append(MLPLayer(in_dim=hid_dim, out_dim=out_dim, bias=bias,
+                                 activation=None, batchnorm=batchnorm,
+                                 dropout=dropout, drop_bef=drop_bef))
```
Comment (mufeili) [inline examples/pytorch/dagnn/main.py:121]: I guess you need to disable dropout for the last layer.

File: examples/pytorch/dagnn/main.py [hunk 15]
```diff
@@ -0,0 +1,272 @@
...
+
+            degs = graph.in_degrees().float().clamp(min=1)
+            norm = torch.pow(degs, -0.5)
+            norm = norm.to(feats.device).unsqueeze(1)
+
+            for _ in range(self.k):
+                feats = feats * norm
+                graph.ndata['h'] = feats
+                graph.update_all(fn.copy_u('h', 'm'),
+                             fn.sum('m', 'h'))
```
Comment (mufeili) [inline examples/pytorch/dagnn/main.py:40]: Seems you need to fix the indent?

File: examples/pytorch/dagnn/main.py [hunk 16]
```diff
@@ -0,0 +1,272 @@
...
+                feats = graph.ndata.pop('h')
```
Comment (mufeili) [inline examples/pytorch/dagnn/main.py:41]: No need to manually pop this. `feats = graph.ndata['h']` is enough.

File: examples/pytorch/dagnn/README.md (hunk 1 above)
Comment (lt610) [reply examples/pytorch/dagnn/README.md]: Now I only leave citation datasets' results.

File: examples/pytorch/dagnn/utils.py (hunk 8 above)
Comment (lt610) [reply examples/pytorch/dagnn/utils.py:20]: If remove this code, the generated seeds will change every time.

File: examples/pytorch/dagnn/utils.py (hunk 9 above)
Comment (lt610) [reply examples/pytorch/dagnn/utils.py:28]: I have added codes to check if cuda is not available.

File: examples/pytorch/dagnn/main.py (hunk 10 above)
Comment (lt610) [reply examples/pytorch/dagnn/main.py]: Sorry, it's redundant. I have removed.

File: examples/pytorch/dagnn/main.py (hunk 14 above)
Comment (lt610) [reply examples/pytorch/dagnn/main.py]: Here dropout locates before transform features.

File: examples/pytorch/dagnn/main.py (hunk 15 above)
Comment (lt610) [reply examples/pytorch/dagnn/main.py:40]: Sorry, If we have added self-loop, it will be redundant.

File: examples/pytorch/dagnn/README.md [hunk 17]
```diff
@@ -0,0 +1,68 @@
...
+--k                int     Number of propagation layers.          Default is 10.
+--hid-dim          int     Hidden layer dimensionalities.         Default is 64.
+--dropout          float   Dropout rate                           Default is 0.8
+```
+
+## Examples
+
+Train a model which follows the original hyperparameters on different datasets.
+```bash
+# Cora:
```
Comment (mufeili) [inline examples/pytorch/dagnn/README.md:55]: Some options seem to be the same for all datasets. In that case, you can simply make them default options.

File: examples/pytorch/dagnn/utils.py (hunk 8 above)
Comment (mufeili) [reply examples/pytorch/dagnn/utils.py:20]: Fine.
